### PR TITLE
L3 filtering support

### PR DESCRIPTION
**Pull request by vlrpl**
What could be interesting to decide here is what the default should be (whether L3 or L2, with L2 being the current default).
ATM, the SKB module has some assumptions that make it skip parsing of packets without the l2 header, making in turn this feature relatively useful. In general, all the hooks that assume the l2 presence would ignore packets that pass the filtering phase.
An additional option was required (to shift the offset) as pcap filters including l2 matches don't return any error from libpcap. Apparently, the filter generation simply ignores primitives that don't apply, and so probing appears to be possible only by parsing the filter (which doesn't seem great, or at least not required as initial approach).

### CODE REVIEW

**Comment by vlrpl** [comment]
> The PR is not looking bad, but I'd like to discuss in depth if we could avoid using cli parameters and make this more user friendly. Given we can know if an skb has l2 or/and l3 information, the ideal implementation would be:

I guess we discussed parts of it offline, the PR's purpose is to find an acceptable way, so agreed, as usual, if we find a better way to consume it, a discussion is welcome.
IMHO, accuracy should come first, then user-friendliness, otherwise things may be confusing for any kind of potential user.

> 
>     * If the filter is L3 only or allows to select based on L3 only (e.g `arp or tcp`): compile it into an L3 filter _and_ an L2 filter, because the skb sometimes only has one of the `mac` and `network` pointers set. This also has the advantage of matching packets in more functions. Note that this could be done in the current PR implementation too.
> 

two filters imply replacing twice, right?

arp should be handled as well?
After a quick test:

```
# this fails
--l3-filter -f arp
Error: libpcap error: expression rejects all packets
# this works
--l3-filter -f "arp or tcp"
```

>     * If the filter is L2 only: compile it into an L2 filter.
> 
> 
> In the BPF side:
> 
> ```
> if any of L2 or L3 filter is set:
>   if skb->mac is set and L2 filter is available:
>     filter using L2
>   else if skb->network is set and L3 filter is available:
>     filter using L3
>   else:
>     reject
> ```

ok, this means to selectively extend (I would not consider it) or replacing twice.

> 
> To determine if a filter is L2 and/or L3 compatible we would need an error from libpcap when parsing it, which could be used to set one or both of the filters. Here is the result of some tests:
> 
>     * L2 specific filters are rejected as expected by libpcap for L3 filters, e.g. `ether multicast`, `ether host xx:xx:xx:xx:xx:xx`, etc. (e.g. "libpcap error: ethernet addresses supported only on ethernet/FDDI/token ring/802.11/ATM LANE/Fibre Channel"), and are accepted  for L2 filters.
> 
>     * Using protocol direct access (`proto[...]`) works and a filter like `ip[6:2] & 0x1fff = 0` matches the same packets if compiled for L2 or for L3. **With the exception of the _ether_ proto which is always accepted and matches from offset 0 regardless of the filter type**.
> 

that's bad, but not that much (if one specifies explicitly the filter type, and is aware of it).

>     * Using `or` statements with a mix of L2 and L3 keywords is not accepted by an L3 filter if it contains an unsupported match, e.g. `ether multicast or tcp` will fail for an L3 filter with "libpcap error: link-layer multicast filters supported only on ethernet/FDDI/token ring/ARCNET/802.11/ATM LANE/Fibre Channel".
> 
>     * Some L2 filters are still accepted and correctly translated to L3 filters, e.g. `ether proto 0x800` does match IPv4 packets only using an L3 filter.
> 
> 
> The above seems to indicate it might be possible to not use an explicit cli flag[1] to select the kind of filter and use the logic described at the start of this comment. With two limitations:
> 
>     * Using the `ether[...]` direct protocol access cannot work because it is ambiguous[2].
> 
>     * Filters of the form `l2 match or l3 match` can't be loaded as L3 filters (aka translated to `l3 match`). Note that we have the same limitation if using cli flags (only the l2 filter could work and be loaded).

but from the user perspective, one would be using l2 matches requesting to shift the offset. I agree that the input validation is lacking, but no match or no load is ok if one uses an explicit flag.
I consider it worse if for some reason a valid L2-only filter doesn't get used in favor of an L3 (that work for some wrong reason), making things way more confusing.

> 
> 
> If the above sounds reasonable, we would probably need to double check the above and make further tests.
> 
> WDYT?
> 
> [1] I think it would still be good to have one, something like `--filter-type`, to force the filter type selection. But that should be a low level parameter. [2] If those cases are only a small subset, it could be detected by us "best effort". E.g. matching `ether[.*]` could be used to not parse a filter as L3.

the idea was to start with the option ("low level" but reliable), and then expand (there's an ongoing major rework).
Simple pattern matching could be enough, but we need to enumerate very carefully the options, otherwise, the risk is that [1] will become the "advanced" way to workaround wrong automatic behavior.

**Comment by vlrpl** [comment]
> > IMHO, accuracy should come first, then user-friendliness, otherwise things may be confusing for any kind of potential user.
> 
> Agreed, we should not make it confusing just for the sake of being easier to use. But idk, I feel both solutions have their drawbacks and potential confusion (due to libpcap not being too strict).
> 

agreed.

> > > ```
> > > * If the filter is L3 only or allows to select based on L3 only (e.g `arp or tcp`): compile it into an L3 filter _and_ an L2 filter, because the skb sometimes only has one of the `mac` and `network` pointers set. This also has the advantage of matching packets in more functions. Note that this could be done in the current PR implementation too.
> > > ```
> > 
> > 
> > two filters imply replacing twice, right?
> 
> Yep.
> 
> > arp should be handled as well? After a quick test:
> > ```
> > # this fails
> > --l3-filter -f arp
> > Error: libpcap error: expression rejects all packets
> > # this works
> > --l3-filter -f "arp or tcp"
> > ```
> 
> Oh, right. So that means some mixed (as in l2+l3) might work for both the L2 and L3 filter. Which is good actually, except if a filter with an _and_ is accepted but did not see an example for this so far.
> 
> > > * Using protocol direct access (`proto[...]`) works and a filter like `ip[6:2] & 0x1fff = 0` matches the same packets if compiled for L2 or for L3. **With the exception of the _ether_ proto which is always accepted and matches from offset 0 regardless of the filter type**.
> > 
> > 
> > that's bad, but not that much (if one specifies explicitly the filter type, and is aware of it).
> 
> Yes, and we have the issue regardless of the solution. The only trick is if we automatically use both filters, but if that is the only problematic construction we could catch it.
> 
> > > * Filters of the form `l2 match or l3 match` can't be loaded as L3 filters (aka translated to `l3 match`). Note that we have the same limitation if using cli flags (only the l2 filter could work and be loaded).
> > 
> > 
> > but from the user perspective, one would be using l2 matches requesting to shift the offset. I agree that the input validation is lacking, but no match or no load is ok if one uses an explicit flag.
> 
> Taking `arp or tcp` as an example:
> 
>     * Using explicit flags: both L2 and L3 would be loaded, L3 filter would not match ARP packets. That's fine IMHO (well, error would be better but not much we can do).
> 
>     * Using automatic selection: both L2 and L3 filters would be used, L2 is preferred when available, match should work as expected (and even match the TCP packets in functions manipulating an skb w/o L2 info). It seems to me this would do the expected and even match more valid packets.
> 
> 
> > I consider it worse if for some reason a valid L2-only filter doesn't get used in favor of an L3 (that work for some wrong reason), making things way more confusing.
> 
> Yes, IMHO that is the only case that needs care. My impression making those tests is only the `ether[..]` statement is an issue as it have a different meaning for L2 and L3. This loos like an historical reason, but of course we should check carefully. Do you see an other problematic one?
> 

not at the moment. TBH, in my tests I went with arp and ether[] and those seemed enough to let me lean for the know-what-you-are-doing approach.
If the problem proves to be confined to that use only, I guess the double filter approach could be fine.
As you said, some tests are needed.

> > > If the above sounds reasonable, we would probably need to double check the above and make further tests.
> > > WDYT?
> > > [1] I think it would still be good to have one, something like `--filter-type`, to force the filter type selection. But that should be a low level parameter. [2] If those cases are only a small subset, it could be detected by us "best effort". E.g. matching `ether[.*]` could be used to not parse a filter as L3.
> > 
> > 
> > the idea was to start with the option ("low level" but reliable), and then expand (there's an ongoing major rework pending). Simple pattern matching could be enough, but we need to enumerate very carefully the options, otherwise, the risk is that [1] will become the "advanced" way to workaround wrong automatic behavior.
> 
> Yep, that's OK to start simple and then extend. I think here we should agree on the way forward as this can impact the simple solution though.

agreed.

**Comment by vlrpl** [comment]
> Thanks @vlrpl, that looks good! There are a few nits, but my main concern is if this is v1.3 material (regarding testing this extensively); WDYT?
> 

I followed a conservative approach for this reason (the approach might lead to a false positive in complex filters but should not generate a false negative).
The best way for this is to reverse the priority (L3 first, L2 as a fallback) as we might want to save cycles, but a no-match in L3 could generate a false negative.

I guess that, if we do not find major problems, this should be ok, but I don't want to "force-push" for it :)
Bugfixes are an option in case we find any.

Let's proceed if we're comfortable

> We should also update the doc about this.
> 

That's right. Some notes are needed. Will do.

> Regardless of the above, I still have to test the PR as I only looked at the code so far.

File: src/collect/collector.rs [hunk 1]
```diff
@@ -134,8 +136,35 @@ impl Collectors {
...
+                        debug!("L3 filter generated.");
+                        f
+                    }
+                }
+            };
+
+            probes.register_filter(Filter::Packet(BpfFilter {
+                key: FilterPacketType::L3,
+                data: fb.to_bytes()?,
+            }))?;
```
[the remark this answers is not in the document]
Comment (vlrpl) [reply src/collect/collector.rs:167]: ether[] it's already there, but in general, all the times we skip L3 we log the reason.
L2 is mandatory, so I found that logging it every time is kind of redundant.
They are all under DEBUG log level, though. Do you suggest moving it to INFO?

File: src/core/filters/packets/filter.rs [hunk 2]
```diff
@@ -35,6 +53,19 @@ impl FilterPacket {
...
     }
 
+    // Generate an empty eBPF filter containing only a single nop
+    // instruction.
+    pub(crate) fn empty_filter() -> Self {
+        let mut ebpf_filter = eBpfProg::new();
+
+        ebpf_filter.add(eBpfInsn::mov32(MovInfo::Imm {
+            dst: BpfReg::R0,
+            imm: 0x1_i32,
```
[the remark this answers is not in the document]
Comment (vlrpl) [reply src/core/filters/packets/filter.rs:63]: yup

File: src/core/probe/manager.rs [hunk 3]
```diff
@@ -179,13 +179,12 @@ impl ProbeManager {
...
-            .filters
-            .iter()
-            .any(|f| std::mem::discriminant(f) == std::mem::discriminant(&filter))
-        {
+        // Avoid duplicate filter types as any Filter but
+        // Filter::Packet variant should be present only once
+        if self.filters.iter().any(|f| {
+            !matches!(f, Filter::Packet(_))
+                && std::mem::discriminant(f) == std::mem::discriminant(&filter)
+        }) {
```
[the remark this answers is not in the document]
Comment (vlrpl) [reply src/core/probe/manager.rs:187]: makes sense. Realistically, this will never happen, but if that happens :)

File: src/core/filters/packets/bpf/include/packet_filter.h [hunk 4]
```diff
@@ -39,6 +42,44 @@ struct retis_packet_filter_ctx {
...
+ * nop frame that will be replaced on load with the actual filtering
+ * instructions. Normally, if no filter gets set, a simple mov r0,
+ * 0x40000 will replace the call. 0x40000 is used as it is also used
+ * by generated cBPF filters, whereas 0 means no match, instead. The
+ * exceeding nops will get removed from the kernel during the load.
+ * If no explicit, nor default filter gets set, call 0xdeadbeef will
+ * fail to load and the verifier will report an error.
+ */
+static __always_inline
+unsigned int packet_filter_mac(struct retis_packet_filter_ctx *ctx)
```
[the remark this answers is not in the document]
Comment (vlrpl) [reply src/core/filters/packets/bpf/include/packet_filter.h:55]: Ok, about renaming. I changed it some times until I got tired and decided to flip the coin and stick with `mac_header`, `network_header` naming.

I had the same mixed feelings about the macro. I decided to keep two separate because 2 didn't seem that bad (although, yes, I don't completely love it).

File: src/core/filters/packets/bpf/include/packet_filter.h [hunk 5]
```diff
@@ -16,6 +16,9 @@ struct retis_packet_filter_ctx {
 	unsigned int ret;	/* outcome of the match (zero if miss). */
 } __attribute__((packed));
 
+/* Lx filter is a nop */
+#define FILTER_NOP	0x1
```
[the remark this answers is not in the document]
Comment (vlrpl) [reply src/core/filters/packets/bpf/include/packet_filter.h:20]: I know you were going to notice this :)
That would be needed if we end up reversing the logic L3 over L2.
In that case, we would know cheaply if L3 missed or was skipped.
It's not needed now, we can remove it if we feel it makes sense.

**Comment by vlrpl** [comment]
> > > Thanks @vlrpl, that looks good! There are a few nits, but my main concern is if this is v1.3 material (regarding testing this extensively); WDYT?
> > 
> > 
> > I followed a conservative approach for this reason (the approach might lead to a false positive in complex filters but should not generate a false negative). The best way for this is to reverse the priority (L3 first, L2 as a fallback) as we might want to save cycles, but a no-match in L3 could generate a false negative.
> 
> That was my impression while reading the code, it looked safe.
> 
> Out of curiosity could you elaborate on the filter inversion and false positive/negative?
> 

The idea of having L3 > L2 sequence is mostly due to cases where you have a valid L3 only filter (e.g. host/port) and the skb has both valid mac_header/network_header.
In that case, assuming L3 check gets performed first, the offset is shifted and the L2 checks are not there. This would be quicker. Looking at the filter assumptions, though, I'm not sure this is exactly what we might end up wanting. For example the simple `host IPv4_ADDR` assumes more than a simple IP {src, dst} match (arp and rarp are taken into account).
So with the inverse order we might lose something (and there's not much we can reliably do about it).

The motivation of the inversion can be fewer instructions to process. But now that we're digging more into it, I realize that there's no big difference (at least in the common cases). Furthermore, given the above, I'm almost convinced that removing the chances of having unexpected behavior is a bigger benefit than skipping some instructions in some cases.

About false negatives, I was considering the explicit cases where alternation allows L3 filters generation but, as a matter of fact, ignoring the L2. We found out "arp or ip" (it seems there are no other cases, but I would not exclude there are more). In the case we have both `network_header` and `mac_header` set, having an L3 filter first means that ARP packets will again be discarded even if are perfectly capable to match them (assuming the logic: "first filter with verdict completes the processing").

TL;DR: there are too many variable not fully known at the moment, and I have the feeling that we will fully enumerate them over time. That being said, I still lean for L2 > L3 even for the future.

> > I guess that, if we do not find major problems, this should be ok, but I don't want to "force-push" for it :) Bugfixes are an option in case we find any.
> 
> That is what I'm leaning forward too (still need to test the PR) but wanted to make sure that was your impression too :)
> 
> Thanks!

thank you!

File: src/core/filters/filters.rs [hunk 6]
```diff
@@ -14,10 +14,13 @@ use crate::core::{
...
 };
 
-use super::meta::filter::FilterMeta;
+use super::{meta::filter::FilterMeta, packets::filter::FilterPacketType};
 
 #[derive(Clone)]
-pub(crate) struct BpfFilter(pub(crate) Vec<u8>);
+pub(crate) struct BpfFilter {
+    pub(crate) key: FilterPacketType,
+    pub(crate) data: Vec<u8>,
```
[the remark this answers is not in the document]
Comment (vlrpl) [reply src/core/filters/filters.rs:22]: do you mean `FilterPacketType(Vec<u8>)`?
If so, `FilterPacketType` with a value is used both as a key in the hashmap used to retrieve the bytes to rewrite and as a selector for the filtering objects that generate L2 or L3 filters (based on the established criteria, like for example [here](https://github.com/retis-org/retis/pull/229/commits/e6d86fc15e7663d8d3ffca11df35198f146a1381#diff-2c71dfc05d3598b49a9c8b44ce12b66811c5754222ca61b95947fdafe63e4e30R141)).
I factored both things and defined this way to use it both as a selector variant and an actual value (needed for the C-related part).

File: src/core/filters/packets/filter.rs [hunk 7]
```diff
@@ -11,18 +11,36 @@ use std::mem;
...
+use crate::core::filters::packets::{
+    cbpf::BpfProg,
+    ebpf::{eBpfProg, BpfReg},
+};
 
 // please keep in sync with FILTER_MAX_INSNS in
 // src/core/probe/kernel/bpf/include/common.h
 const FILTER_MAX_INSNS: usize = 4096;
 
+#[allow(dead_code)]
```
[the remark this answers is not in the document]
Comment (vlrpl) [reply src/core/filters/packets/filter.rs:25]: I removed in the next patch where it gets actually used

File: src/collect/collector.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (vlrpl) [reply src/collect/collector.rs:167]: IMHO, if we keep it as `DEBUG` it should be functional for debugging, and avoiding L2 is fine (because it `bail!()` in that case).
If we want `INFO` we should probably go with `L2[+L3] installed`. I tend to prefer `debug!()` but I have no strong preference. Whatever you prefer is fine.

**Comment by vlrpl** [comment]
doc update missing. Will send new version soon.

File: src/core/filters/filters.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (vlrpl) [reply src/core/filters/filters.rs:22]: > > do you mean `FilterPacketType(Vec<u8>)`?
> 
> At first I meant:
> 
> ```rust
> enum BpfFilter {
>     L2(Vec<u8>),
>     L3(Vec<u8>),
> }
> ```

I wrote it a bit quick&dirty, but I meant something like that with `FilterPacketType` being `L2` or `L3` variant.

> 
> Or maybe directly in `enum Filter`; but then I though: shouldn't the L2 and the L3 filter be part of the same object, as they make up the `Filter::Packet`. It seems logical to have them both (L3 could be in an Option<>) in one `BpfFilter` object, WDYT?
> 

BpfFilter was initially supposed to be some kind of generic object containing the binary to consume (e.g. one could have been freplaced when we used to dynamic relink filters, and another inlined). Things changed and reusing it for inlining still made sense. Now things are changing again as it happens.

This makes sense (keeps the same intent for `BpfFilter`, assuming it still makes sense) but has the drawback of allowing duplicates while registering (not sure it is bad):
```
pub(crate) enum Filter {
    Packet(L2, BpfFilter), // `get_type()`, `get_data()`
    Packet(L3, BpfFilter), // `get_type()`, `get_data()`
    Meta(FilterMeta),
}
```
```
pub(crate) enum Filter {
    Packet(BpfFilter, Option<BpfFilter>), // `.get_l2()`, `.get_l3()`
    Meta(FilterMeta),
}
```

in this second case we should still keep FilterPacketType::{L2 = magic_l2, L3 = magic_l3} (or they can be simply constants) and we'd probably need some `.get_magic_l{2,3}()`  to perform:
```
if let Filter::Packet(bpf_filter) = filter {
    filters::register_filter(bpf_filter.key as u32, filter)?;
}
```

in `setup_dyn_filters()`

This would not allow duplicates in the `filters` vector (no need).
Some changes are required, maybe more for the second approach.

> (No sure how easy or realistic that would be).

[the remark this answers is not in the document]
Comment (vlrpl) [reply src/core/filters/filters.rs:22]: > I had to try my proposal to see how it would go, https://paste.ack.tf/e9caa6@raw (untested). I agree, the link between `Filter` and `FilterPacketType` has to be done manually in this version, which isn't 100% perfect.
> 
> I'm starting to think the first proposal would be best too.

of course, map at impl level or match/map where needed. Getter was to give an idea.
I think I'm losing track of first :) 
First of what discussion?

[the remark this answers is not in the document]
Comment (vlrpl) [reply src/core/filters/filters.rs:22]: > I hope same as the one you were referencing "Slight preference for the first" :)
> 
> ```rust
> pub(crate) enum Filter {
>     Packet(L2, BpfFilter),
>     Packet(L3, BpfFilter),
>     Meta(FilterMeta),
> }
> ```

Yes, that is it.
Just checking as one never knows :)
I think the only non-ideal thing here is the exception in filters, but that's confined to that and after all, reflects the nature of multiple packet filters vs other unique ones.

File: src/collect/collector.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (vlrpl) [reply src/collect/collector.rs:167]: SGTM, I'm convinced that it's probably better to make "automatic logic" stand out. The user would align the expectations in this way. The reason (debug) is useful anyway, mostly for us to quickly understand what's going on.

While we're speaking, could make sense to dump `mac_header` and `network_header` in the skb module?
Maybe as part of a debug section?
Or consider some raw reporting of `bpf_snprintf_btf()`, e.g. tackling the local trace pipe and making that info available at some log level.
